test/test_ops.py

```diff
@@ -1267,10 +1267,73 @@ def test_double_slice(self):
+    self.helper_test_exception([(3,3,3)], lambda x: torch.nn.functional.pad(x, (2,0,2)), lambda x: x.pad2d((2,0,2)),
+                               expected=(RuntimeError, ValueError))
+    # raise error for negative output size
+    self.helper_test_exception([(3,3,3)], lambda x: torch.nn.functional.pad(x, (0, -7)), lambda x: x.pad2d((0, -7)),
+                               expected=(RuntimeError, AssertionError))
+    # raise error for mode string typo
+    self.helper_test_exception([(3,3,3)], lambda x: torch.nn.functional.pad(x, (3,0), mode="typo"), lambda x: x.pad2d((3,0), mode="typo"),
+                               expected=(NotImplementedError, ValueError))
+
+  def test_pad2d_modes(self):
+    # basic pads
+    for shp, pad in (((1,1,5,5), (0,2,3,2)), ((5,5,5), (0,2)),
+                    ((1,1,5,5,5), (1,2,3,4,1,2)), ((1,1,5,5,5), (1,2,3,4,1,2))):
+      for mode in ("reflect", "replicate", "circular"):
+        with self.subTest(shp=shp, pad=pad, mode=mode):
+          helper_test_op([shp], lambda x: torch.nn.functional.pad(x, pad, mode=mode), lambda x: x.pad2d(pad, mode=mode))
+
+    # replicate large pads
+    helper_test_op([(1,1,5,5)], lambda x: torch.nn.functional.pad(x, (3,11,0,30), mode="replicate"), lambda x: x.pad2d((3,11,0,30), mode="replicate"))
+
+    # wrap around exactly once
+    helper_test_op([(1,1,5,5)], lambda x: torch.nn.functional.pad(x, (5,5,0,5), mode="circular"), lambda x:x.pad2d((5,5,0,5),mode="circular"))
+    # raise error when trying to wrap around more than once
+    self.helper_test_exception([(1,1,5,5)],
+                                lambda x: torch.nn.functional.pad(x, (3,6,0,0),mode="circular"), lambda x: x.pad2d((3,6,0,0),mode="circular"),
+                                expected=(RuntimeError, ValueError))
+    # reflect exactly once
+    helper_test_op([(1,1,5,5)], lambda x: torch.nn.functional.pad(x, (4,4,0,4), mode="reflect"), lambda x:x.pad2d((4,4,0,4),mode="reflect"))
+    # raise error for relfection padding when pad >= input size
+    self.helper_test_exception([(1,1,5,5)],
+                                lambda x: torch.nn.functional.pad(x, (3,5,0,0),mode="reflect"), lambda x: x.pad2d((3,5,0,0),mode="reflect"),
+                                expected=(RuntimeError, ValueError))
+
+  @unittest.skipIf(CI, "non-constant modes with negative pads have odd numerical errors in torch CI")
+  def test_pad2d_modes_negative_pads(self):
+    # negative pads
+    for shp, pad in (((3,3,3,3), (-1,2,2,-1)), ((1,1,5,5), (3,-3,0,-3)), ((1,1,5,5), (3,-5,1,-5)), ((1,1,5,5), (0,0,0,-5))):
+      for mode in ("reflect", "replicate", "circular"):
+        with self.subTest(shp=shp, pad=pad, mode=mode):
+          helper_test_op([shp], lambda x: torch.nn.functional.pad(x, pad, mode=mode), lambda x: x.pad2d(pad, mode=mode))
+
+    # TODO: fix backward for circular negative pads, problems occur when more-than-once wrap around happens when negative pads come before
```

Comment:
these negative pads with modes is cursed
for example for forwards:
![image](https://github.com/user-attachments/assets/c4568a88-8abb-44dc-a1d8-bc57cb14c050)
I have no idea why torch does this.
I was very close to slapping a `if any(pB<0 or pA<0 for pB,pA in pX): raise NotImplementedError('negative padding with circular mode not implemented')` on it and calling it a day lol

but for the rest I think our pad2d matches torch's pad.